File: ch4/clip8/boj1915/main.py
import sys

n, m = map(int, sys.stdin.readline().split())
arr = []
dp = [[0] * (m + 1) for _ in range(n + 1)]

for _ in range(n):
    row = list(map(int, list(sys.stdin.readline().strip())))
    arr.append(row)

max_size = 0

for i in range(1, n + 1):
    for j in range(1, m + 1):
        if arr[i - 1][j - 1] == 1:
            dp[i][j] = min(dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1]) + 1
            max_size = max(max_size, dp[i][j])

print(max_size * max_size)

File: ch4/clip8/boj1915/main.cpp
#include <iostream>

int main()
{
    int n, m, board;
    int maxSize[1001][1001];
    int result = 0;

    scanf("%d %d", &n, &m);

    for(int j = 0; j <= n; j++)
    {
        maxSize[j][0] = 0;
    }

    for(int i = 0; i <= m; i++)
    {
        maxSize[0][i] = 0;
    }

    for(int j = 1; j <= n; j++)
    {
        for(int i = 1; i <= m; i++)
        {
            scanf("%1d", &board);
            if(1 == board)
            {
                maxSize[j][i] = maxSize[j - 1][i - 1];
                if(maxSize[j][i] > std::min(maxSize[j - 1][i], maxSize[j][i - 1]))
                {
                    maxSize[j][i] = std::min(maxSize[j - 1][i], maxSize[j][i - 1]);
                }
                maxSize[j][i]++;

                if(maxSize[j][i] > result)
                {
                    result = maxSize[j][i];
                }
            }
            else
            {
                maxSize[j][i] = 0;
            }
        }
    }

    printf("%d", result * result);

    return 0;
}